compiler/rustc_parse/src/parser/item.rs

```diff
@@ -2332,10 +2332,20 @@ impl<'a> Parser<'a> {
                 }
             }
         };
+
+        // Store the end of function parameters to give better diagnostics
+        // inside `parse_fn_body()`.
+        let fn_params_end = self.prev_token.span.shrink_to_hi();
+
         generics.where_clause = self.parse_where_clause()?; // `where T: Ord`
 
+        // `fn_params_end` is needed only when it's followed by a where clause.
+        let fn_params_end =
+            if generics.where_clause.has_where_token { Some(fn_params_end) } else { None };
```

Comment:
what's the harm of always passing in that span?

Reply:
Checking if a where clause exists makes sure we don't incorrectly suggest to place the second return type after function parameters for code like `fn fun<T>() -> u8 -> u8 {}`.